core/arch/arm/kernel/misc_a32.S

```diff
@@ -41,20 +41,63 @@ FUNC get_core_pos , :
 	bx	lr
 END_FUNC get_core_pos
 
-/* uint32_t read_usr_sp(void) */
-FUNC read_usr_sp , :
-	mrs	r1, cpsr
+LOCAL_FUNC temp_set_mode , :
+
 	cps     #CPSR_MODE_SYS
+	cmp	r0, #CPSR_MODE_SYS
+	moveq	pc, r2
+	cmp	r0, #CPSR_MODE_USR
+	moveq	pc, r2
+
+	cps     #CPSR_MODE_FIQ
+	cmp	r0, #CPSR_MODE_FIQ
+	moveq	pc, r2
+
+	cps     #CPSR_MODE_IRQ
+	cmp	r0, #CPSR_MODE_IRQ
+	moveq	pc, r2
+
+	cps     #CPSR_MODE_SVC
+	cmp	r0, #CPSR_MODE_SVC
+	moveq	pc, r2
+
+	cps     #CPSR_MODE_MON
+	cmp	r0, #CPSR_MODE_MON
+	moveq	pc, r2
+
+	cps     #CPSR_MODE_ABT
+	cmp	r0, #CPSR_MODE_ABT
+	moveq	pc, r2
+
+	cps     #CPSR_MODE_UND
+	cmp	r0, #CPSR_MODE_UND
+	moveq	pc, r2
+
+	b .
+
+END_FUNC temp_set_mode
+
+/* uint32_t read_mode_sp(int cpu_mode) */
+FUNC read_mode_sp , :
+	ldr	r2, =_sp_ret
```

Comment:
This is clumsy, I would prefer an implementation where `temp_set_mode()` switched the mode (by calculating the new cpsr) and returned the original mode to restore in `r0`. Please use `bx` instead of `b`, you can push original `lr` on the stack before calling `temp_set_mode()`.